src/borg/key.py

```diff
@@ -57,6 +59,80 @@ class RepoKeyNotFoundError(Error):
     """No key entry found in the config of repository {}."""
 
 
+class SignatureError(Error):
+    """Invalid signature for {}."""
+
+
+class SignedFile(FileLikeWrapper):
+    # generic enough that it can be used without a KeyBase
+    def __init__(self, key, path, write):
+        self.path = path
+        self.writing = write
+        mode = 'wb' if write else 'rb'
+        self.file_fd = open(path, mode)
+        self.signer = StreamSigner_HMAC_SHA512(key, self.file_fd, write)
+        self.fd = self.signer  # for FileLikeWrapper
+        self.sign_filename()
+        if write:
+            self.signatures = {}
+        else:
+            self.signatures = self.read_signatures(path, self.signer)
+        self.failed = False
+
+    def sign_filename(self):
+        # Sign the name of the file as well, but only the basename, ie. not the path. In Borg
+        # the name itself encodes the context (eg. index.N, cache, files), while the path doesn't matter,
+        # and moving eg. a repository or cache directory is supported.
+        # Changing the name however imbues a change of context that is not permissible.
+        filename = os.path.basename(self.path)
+        self.signer.update(b'%d' % len(filename))
+        self.signer.update(filename.encode())
+
+    @staticmethod
+    def read_signatures(path, signer):
+        signature_path = path + '.bsi'
+        try:
+            with open(signature_path, 'r') as fd:
+                signature = json.load(fd)
+                # Provisions for agility now, implementation later, but make sure the on-disk joint is oiled.
+                algorithm = signature['algorithm']
```

Comment:
These are caught below (OSError, ValueError, TypeError, LookupError); I can change the latter into KeyError (instead of LookupError - I thought indexing a list with a string would be IndexError, but it's actually TypeError).